octree_web_viewer/src/backend.rs

```
@@ -65,82 +73,92 @@ fn pad(input: &mut Vec<u8>) {
-        Ok(Response::with((content_type, iron::status::Ok, reply_blob)))
+impl<S: 'static> Handler<S> for NodesData {
+    type Result = FutureResponse<HttpResponse>; //alias forBox<Future<Item=HttpResponse, Error=Error>>;
+
+    fn handle(&self, req: &HttpRequest<S>) -> Self::Result {
+        let octree = Arc::clone(&self.octree); //has to be moved into future
+        let message_body_future = Json::<Vec<String>>::extract(req).from_err();
+        message_body_future
+            .and_then(move |extract_result| {
+                let data: Vec<String> = Json::into_inner(extract_result);
+                let nodes_to_load = data
+                    .into_iter()
+                    .map(|e| octree::NodeId::from_str(e.as_str()));
+
+                // So this is godawful: We need to get data to the GPU without JavaScript herp-derping with
+                // it - because that will stall interaction. The straight forward approach would be to ship
+                // json with base64 encoded values - unfortunately base64 decoding in JavaScript yields a
+                // string which cannot be used as a buffer. So we would need to manually convert this into
+                // an Array with is very slow.
+                // The alternative is to binary encode the whole request and parse it on the client side,
+                // which requires careful constructing on the server and parsing on the client.
+                let start = time::precise_time_ns(); //timing is not comparable to the previous implementation in iron
+                let mut reply_blob = Vec::<u8>::new();
+
+                let mut num_nodes_fetched = 0;
+                let mut num_points = 0;
+                for node_id in nodes_to_load {
+                    let mut node_data = match octree.get_node_data(&node_id) {
+                        Ok(n) => n,
+                        Err(err) => panic!("Could not get node {}: {} ", node_id, err),
```

Can we instead propagate this error to the outside?

---

done!